### PR TITLE
Improvement proposal

### PR DESCRIPTION
**Pull request by perrettecl**
Improvements:
1. request interface do not force the copy of the message => malamute & unix socket need a copy, but maybe not all of the implementation. If the copy is needed the implementation must do the copy.

2. Add todo arround mlm_sendto function: We need to catch error there.

3. Remove detach thread for execution of the listener: If the user of the lib need an to have the listen running in different thread he can choose his own strategy (detach, pull of thread....).

Signed-off-by: Clement Perrette <clementperrette@eaton.com>

### CODE REVIEW

File: src/fty_common_messagebus_malamute.cc [hunk 1]
```diff
@@ -313,11 +323,17 @@ namespace messagebus {
         if( syncResponse == false ) {
             auto iterator = m_subscriptions.find (subject);
             if (iterator != m_subscriptions.end ()) {
-                //iterator->second(msg);
-                std::thread (iterator->second, msg).detach();
+                try {
+                    (iterator->second)(msg);
+                }
```
Comment (boricj) [inline src/fty_common_messagebus_malamute.cc:331]: The reason why we spawned another thread is because we don't want to block message reception if the client has a long-duration callback (for example, making a synchronous request). I'm not a fan of turning this part into a potential huge serialization bottleneck. What we can do alternatively is a worker pool in order not to spawn a thread on each received message.

What do other libraries do in this situation?

File: src/fty_common_messagebus_malamute.cc [hunk 2]
```diff
@@ -330,8 +346,15 @@ namespace messagebus {
 
         auto iterator = m_subscriptions.find (subject);
         if (iterator != m_subscriptions.end ()) {
-            //iterator->second(msg);
-            std::thread (iterator->second, msg).detach();
+                try {
+                    (iterator->second)(msg);
+                }
```
Comment (boricj) [inline src/fty_common_messagebus_malamute.cc:354]: Same as above.

**Review by boricj — CHANGES_REQUESTED** [review]
Need clarification/decision on library threading model.

File: src/fty_common_messagebus_malamute.cc (hunk 1 above)
Comment (perrettecl) [reply src/fty_common_messagebus_malamute.cc:331]: For me detach is the bad model. if we have long duration callback, the client need to do the detach in his call back.

File: src/fty_common_messagebus_malamute.cc [hunk 3]
```diff
@@ -222,10 +226,16 @@ namespace messagebus {
...
         std::unique_lock<std::mutex> lock(m_cv_mtx);
-        message.metaData().emplace(Message::REPLY_TO, m_clientName);
-        zmsg_t *msg = _toZmsg (message);
-        mlm_client_sendto (m_client, iterator->second.c_str(), requestQueue.c_str(), nullptr, 200, &msg);
+        msg.metaData().emplace(Message::REPLY_TO, m_clientName);
+        zmsg_t *msgMlm = _toZmsg (msg);
+
+        //Todo: Check error code after sendto
+        mlm_client_sendto (m_client, iterator->second.c_str(), requestQueue.c_str(), nullptr, 200, &msgMlm);
+
```
Comment (NavarroPierre) [inline src/fty_common_messagebus_malamute.cc:241]: Didn't found the return code of this function in mlm_client.h.... maybe
if( mlm_client_sendto(....) != 0 ) {
    throw MessageBusException("Request failure");
}

**Review by NavarroPierre — APPROVED** [review]

**Review by boricj — APPROVED** [review]
